File: trees/c-or-cpp/pre-in-post-traversal.cpp
#include<iostream>
#define endl "\n"
using namespace std;


// Basic Structure of a Tree
struct Node{
    int data;
    struct Node *left;
    struct Node *right;

    Node(int val){
        data = val;
        left = NULL;
        right = NULL;
    }
};

/* Pre, In and Post order Traversal of a Binary Tree using recusrion */
/* Complexity for each Traversal : O(n) */

// LST -> Left SubTree   RST -> Right SubTree

// Pre-Order goes like :-   Root, LST, RST (Recursively)
void pre_order(struct Node* root){
    if(root == NULL)
      return;
    cout<<root->data<<" ";
    pre_order(root->left);
    pre_order(root->right);
}

// In-Order goes like :-   LST, Root, RST (Recursively)
void in_order(struct Node* root){
    if(root == NULL)
      return;
    in_order(root->left);
    cout<<root->data<<" ";
    in_order(root->right);
}

// Pre-Order goes like :-  LST, RST, Root (Recursively)
void post_order(struct Node* root){
    if(root == NULL)
      return;
    post_order(root->left);
    post_order(root->right);
    cout<<root->data<<" ";
}


int main(){
    Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    root->right->right = new Node(7);

    /*                       
                    1
                   / \
                  2   3
                 / \   \
                4   5   7    
        Expected-> pre-order :  1,2,4,5,3,7
                   in-order  :  4,2,5,1,3,7
                   post-order:  4,5,2,7,3,1
    
    */
   cout<<"pre-order : "; pre_order(root); cout<<endl;
   cout<<"in-order : "; in_order(root); cout<<endl;
   cout<<"post-order : "; post_order(root); cout<<endl;

    return 0;
}

File: algorithms/Java/trees/pre-in-post-traversal.java
class Node{  
    int value;  
    Node left;  
    Node right;  

    public Node(int value) {  
        this.value = value;  
    }  
}

public class pre_in_post_traversal{

    public static void preOrder(Node node){
        if (node == null) return;

        System.out.print(node.value + ", ");
        preOrder(node.left);
        preOrder(node.right);
    }

    public static void inOrder(Node node){
        if(node == null) return;

        inOrder(node.left);
        System.out.print(node.value + ", ");
        inOrder(node.right);
    }

    public static void postOrder(Node node){
        if(node == null) return;
        postOrder(node.left);
        postOrder(node.right);
        System.out.print(node.value + ", ");
    }


    public static void main(String [] args){
        Node root = new Node(4);
        root.left = new Node(2);
        root.right = new Node(7);
        root.left.right = new Node(6);
        root.left.left = new Node(1);
        root.right.right = new Node(9);

        System.out.print("pre order: ");
        preOrder(root);
        
        System.out.println("\n");
        System.out.print("in order: ");
        inOrder(root);

        System.out.print("\n\npost order: ");
        postOrder(root);
    }
}

/*
tree structure:
        4
      /   \
    2       7
  /   \       \
1      6       9

to run the file:
    javac .\pre_in_post_traversal.java
    java pre_in_post_traversal

results:
    pre order: 4, 2, 1, 6, 7, 9, 

    in order: 1, 2, 6, 4, 7, 9,

    post order: 1, 6, 2, 9, 7, 4,

*/
